x/valset-pref/validator_set.go

```diff
@@ -183,21 +182,25 @@ func (k Keeper) PreformRedelegation(ctx sdk.Context, delegator sdk.AccAddress, e
 	for _, diff_val := range diffValSet {
 		for diff_val.amount.GT(sdk.NewDec(0)) {
 			source_val := diff_val.valAddr
-			// FindMin returns the index and MinAmt of the minimum amount in diffValSet
-			target_val, idx := k.FindMin(diffValSet)
+			target_val, idx := k.FindMin(diffValSet, source_val)
+
+			// checks if there are any more redelegation possible
+			if target_val.amount.Equal(sdk.NewDec(0)) {
+				break
+			}
 
 			validator_source, err := sdk.ValAddressFromBech32(source_val)
 			if err != nil {
-				return fmt.Errorf("validator address not formatted")
+				return fmt.Errorf("source validator address not formatted")
 			}
 
 			validator_target, err := sdk.ValAddressFromBech32(target_val.valAddr)
 			if err != nil {
-				return fmt.Errorf("validator address not formatted")
+				return fmt.Errorf("destination validator address not formatted")
 			}
 
 			// reDelegationAmt to is the amount to redelegate, which is the min of diffAmount and target_validator
-			reDelegationAmt := sdk.MinDec(target_val.amount.Abs(), diff_val.amount)
+			reDelegationAmt := sdk.MinDec(target_val.amount.Abs(), diff_val.amount).TruncateDec()
```

Comment:
If we truncateDec here, would it not be possible to go into a infinite for loop?

Reply:
hrmm i haven't seen any cases for  infinite loop yet. I will be simulator testing this feature so, if i find anything i will address it